roxie/udplib/udpsha.cpp

```diff
@@ -91,39 +91,58 @@ queue_t::~queue_t()
     }
 }
 
-unsigned queue_t::available()
+void queue_t::interrupt()
 {
-    CriticalBlock b(c_region);
-    if (count < limit)
-        return limit - count;
-    return 0;
+    data_avail.interrupt();
 }
 
-int queue_t::free_slots() 
+void queue_t::doEnqueue(DataBuffer *buf)
 {
-    int res=0;
-    while (res <= 0)
+    // Must currently be called within a critical section.  Does not signal - that should be done outside the CS.
+    // Could probably be done lock-free, which given one thread using this is high priority might avoid some
+    // potential priority-inversion issues. Or we might consider using PI-aware futexes here?
+    if (tail)
+    {
+        assert(head);
+        assert(!tail->msgNext);
+        tail->msgNext = buf;
+    }
+    else
+    {
+        assert(!head);
+        head = buf;
+    }
+    tail = buf;
+    count.fastAdd(1); // inside a critical section, so no need for atomic inc.
+#ifdef _DEBUG
+    if (count > limit)
+        DBGLOG("queue_t::pushOwn set count to %u", count.load());
+#endif
+}
+
+void queue_t::pushOwnWait(DataBuffer * buf)
+{
+    assert(!buf->msgNext);
+    for (;;)
     {
-        c_region.enter();
-        res = limit - count;
-        if (res <= 0)
-            signal_free_sl++;
-        c_region.leave();
-        if (res <= 0)
         {
-            while (!free_sl.wait(3000))
+            CriticalBlock b(c_region);
+
+            if (count < limit)
```

Comment:
I wonder if you could check this before grabbing the critsec, and then either check again after or accept that just occasionally you would get an incorrect answer if count was being updated by another thread - would it matter?

Not sure it would make a lot of difference, depends on whether it's normal for the queue to be full when first entering.

Reply:
I think if true the doEnqueue() needs to be inside the critical section, and if false he signal_free_sl needs to be inside the cs, so there wouldn't be any benefit.